clippy_lints/src/unused_io_amount.rs

```diff
@@ -45,126 +46,241 @@ declare_clippy_lint! {
-            is_trait_method(cx, call, sym::IoWrite)
-        };
-
-        match (read_trait, write_trait, symbol, is_await) {
-            (true, _, "read", false) => span_lint_and_help(
-                cx,
-                UNUSED_IO_AMOUNT,
-                expr.span,
-                "read amount is not handled",
-                None,
-                "use `Read::read_exact` instead, or handle partial reads",
-            ),
-            (true, _, "read", true) => span_lint_and_help(
-                cx,
-                UNUSED_IO_AMOUNT,
-                expr.span,
-                "read amount is not handled",
-                None,
-                "use `AsyncReadExt::read_exact` instead, or handle partial reads",
-            ),
-            (true, _, "read_vectored", _) => {
-                span_lint(cx, UNUSED_IO_AMOUNT, expr.span, "read amount is not handled");
-            },
-            (_, true, "write", false) => span_lint_and_help(
-                cx,
-                UNUSED_IO_AMOUNT,
-                expr.span,
-                "written amount is not handled",
-                None,
-                "use `Write::write_all` instead, or handle partial writes",
-            ),
-            (_, true, "write", true) => span_lint_and_help(
-                cx,
-                UNUSED_IO_AMOUNT,
-                expr.span,
-                "written amount is not handled",
-                None,
-                "use `AsyncWriteExt::write_all` instead, or handle partial writes",
-            ),
-            (_, true, "write_vectored", _) => {
-                span_lint(cx, UNUSED_IO_AMOUNT, expr.span, "written amount is not handled");
-            },
-            _ => (),
+fn emit_lint(cx: &LateContext<'_>, span: Span, op: IoOp, wild_cards: &[Span]) {
+    let (msg, help) = match op {
+        IoOp::AsyncRead(false) => (
+            "read amount is not handled",
+            Some("use `AsyncReadExt::read_exact` instead, or handle partial reads"),
+        ),
+        IoOp::SyncRead(false) => (
+            "read amount is not handled",
+            Some("use `Read::read_exact` instead, or handle partial reads"),
+        ),
+        IoOp::SyncWrite(false) => (
+            "written amount is not handled",
+            Some("use `Write::write_all` instead, or handle partial writes"),
+        ),
+        IoOp::AsyncWrite(false) => (
+            "written amount is not handled",
+            Some("use `AsyncWriteExt::write_all` instead, or handle partial writes"),
+        ),
+        IoOp::SyncRead(true) | IoOp::AsyncRead(true) => ("read amount is not handled", None),
+        IoOp::SyncWrite(true) | IoOp::AsyncWrite(true) => ("written amount is not handled", None),
+    };
+
+    span_lint_and_then(cx, UNUSED_IO_AMOUNT, span, msg, |diag| {
+        if let Some(help_str) = help {
+            diag.help(help_str);
         }
-    }
+        for span in wild_cards {
+            diag.span_note(*span, "io bytes are ignored");
```

Comment:
Hmmm I'm not sure if this should be "io", "IO", or "I/O".

Reply:
I prefer `I/O` to be honest 🤔

Reply:
The problem is that anything capitalized breaks guidelines 😅

Reply:
I'll talk about this in Zulip. The [style guide](https://rustc-dev-guide.rust-lang.org/diagnostics.html?highlight=diagnostic#diagnostic-output-style-guide) does indeed say:

> - Error, Warning, Note, and Help messages start with a lowercase letter and do not end with punctuation.

But, "I/O" is an acronym... And using `io` would break the first style convention...

Reply:
It seems like there's an unanimously support for "I/O". Could you change it =^w^=

Reply:
Of course! what should I do about the style check :D ?

Reply:
? What style check? Like, the `x test tidy` in upstream Rustc?

Reply:
running `cargo test` with I/O results in the test suite failing because of the style check on I/O. I could reword the note if that's okay with you.

Reply:
Yep, reword it if that's happening. I've tried using `cargo test` in this PR, and apart from having to bless the `.stderr` files and rename the `//~ NOTE:` annotations, everything went just fine. ¯ \\ \_ (ツ) \_ / ¯

Reply:
What do you think of this:

```rust
match s.write(b"test") {
    //~^ ERROR: written amount is not handled
    Ok(_) => todo!(),
    //~^ NOTE: the result is consumed here, but the amount of I/O bytes remains unhandled
    Err(_) => todo!(),
}
```